tvm/src/codegen/build_util.cc

```diff
@@ -325,351 +327,159 @@ void PrintCopyBack(TVMArray* arr,
+    auto next = test_file.find('\n', pos);
+    test_file.insert(next + 1, "#include \"kernel.h\"\n");
+
+    // create typedef list 
+    std::unordered_map<std::string, std::string> typedef_map({ 
+        { "ap_uint<32>" , "ubit32" }, 
+        { "ap_int<32>"  , "bit32"  } 
+    });
+
+    for (auto& kv : typedef_map) {
+      while (test_file.find(kv.first) != std::string::npos)
+        test_file.replace(test_file.find(kv.first), 
+            kv.first.length(), kv.second);
     }
-  }
 
-  // print top func
-  stream << "\n";
-  PrintIndent(stream, indent);
-  stream << "top( ";
-  for (size_t i = 0;i < arg_stream_types.size(); i++) {
-    if (i != arg_stream_types.size() - 1){
-      stream << "source_wrapper_temp_" << i;
-      stream << ", ";
-    } else {
-      stream << "source_wrapper_temp_" << i;
-      stream << ");\n";
+    // generate header file
+    std::ofstream header;
+    header.open("__tmp__/kernel.h");
+    header << "#ifndef __KERNEL_H__\n" 
+           << "#define __KERNEL_H__\n\n";
+    header << "#include <ap_int.h>\n";
+    header << "#include <ap_fixed.h>\n";
+    header << "#include <hls_stream.h>\n";
+    for (auto& kv : typedef_map) {
+      header << "typedef " << kv.first << " "
+             << kv.second << ";\n";
     }
 
-  }
-  stream << "\n";
+    // locate top function
+    CHECK(test_file.find("test(") != std::string::npos) 
+      << "cannot find top function";
+    size_t dut = test_file.find("test(");
+    size_t begin = test_file.rfind('\n', dut);
+    size_t end = test_file.find(')', dut) + 1;
 
-  // read back return val
-  for (int k = arg_stream_types.size() - 1; 
-       k > args.size() - 2; k--) {
-    auto shape = std::get<3>(arg_stream_types[k]);
-    for (size_t i = 0; i < shape.size(); i++) {
-      PrintIndent(stream, indent);
-      stream << "for (int i" << i << " = 0; ";
-      stream << "i" << i << " < " << shape[i] <<  "; ";
-      stream << "i" << i << "++) {\n";
-      indent += 2;
-    
-      if (i == shape.size() - 1) {
-        PrintIndent(stream, indent);
-        stream << "source_wrapper_" << k;
-        stream << "[i" << shape.size() - 1;
-        int mul = 1;
-        for (size_t j = shape.size() - 1; j > 0; j--) {
-          mul *= shape[j];
-          stream << " + i" << j - 1 << "*" << mul;
-        }
-        stream << " ] = ";
-    
-        stream << "source_wrapper_temp_" << k;
-        for (size_t j = 0; j < shape.size(); j++) {
-          stream << "[i" << j << "]";
-        }
-        stream <<";\n";
-      }
-    }
-    for (size_t i = 0;i < shape.size(); i++) {
-        indent -= 2;
-        PrintIndent(stream, indent);
-        stream << "}\n";
-    }
+    header << test_file.substr(begin, end - begin) 
+           << ";\n" << "\n#endif";
+    header.close();
+    stream << "#include <ap_int.h>\n";
+    stream << "#include <ap_fixed.h>\n";
+
+  } else if (platform == "aocl")  {
+    // stream << "#include \"ihc_apint.h\"\n";
```

Comment:
Is this needed or not?

Reply:
It's needed. But to include this header I need to add an additional flag in Makefile.